pkg/proxy/ipvs/netlink_linux.go

```diff
@@ -143,52 +148,62 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
 // 10.0.0.10  scope host  src 10.0.0.10
 // Then cut the unique src IP fields,
 // --> result set: [10.0.0.1, 10.0.0.10]
-
-// If filterDev is specified, the result will discard route of specified device and cut src from other routes.
 func (h *netlinkHandle) GetLocalAddresses(dev, filterDev string) (sets.String, error) {
-	chosenLinkIndex, filterLinkIndex := -1, -1
 	if dev != "" {
-		link, err := h.LinkByName(dev)
-		if err != nil {
-			return nil, fmt.Errorf("error get device %s, err: %v", dev, err)
+		if filterDev != "" {
+			return nil, fmt.Errorf("Invalid argument")
 		}
-		chosenLinkIndex = link.Attrs().Index
-	} else if filterDev != "" {
-		link, err := h.LinkByName(filterDev)
+		ifi, err := net.InterfaceByName(dev)
 		if err != nil {
-			return nil, fmt.Errorf("error get filter device %s, err: %v", filterDev, err)
+			return nil, fmt.Errorf("Could not get interface %s: %v", dev, err)
 		}
-		filterLinkIndex = link.Attrs().Index
-	}
-
-	routeFilter := &netlink.Route{
-		Table:    unix.RT_TABLE_LOCAL,
-		Type:     unix.RTN_LOCAL,
-		Protocol: unix.RTPROT_KERNEL,
+		ips := sets.NewString()
+		h.appendAddresses(ips, ifi)
+		return ips, nil
 	}
-	filterMask := netlink.RT_FILTER_TABLE | netlink.RT_FILTER_TYPE | netlink.RT_FILTER_PROTOCOL
 
-	// find chosen device
-	if chosenLinkIndex != -1 {
-		routeFilter.LinkIndex = chosenLinkIndex
-		filterMask |= netlink.RT_FILTER_OIF
+	if filterDev == "" {
+		return nil, fmt.Errorf("Invalid argument")
 	}
-	routes, err := h.RouteListFiltered(netlink.FAMILY_ALL, routeFilter, filterMask)
+	ifaces, err := net.Interfaces()
 	if err != nil {
-		return nil, fmt.Errorf("error list route table, err: %v", err)
+		return nil, fmt.Errorf("Could not get interfaces: %v", err)
+	}
+	ips := sets.NewString()
+	for _, i := range ifaces {
+		if i.Name == filterDev {
+			continue
+		}
+		h.appendAddresses(ips, &i)
 	}
-	res := sets.NewString()
-	for _, route := range routes {
-		if route.LinkIndex == filterLinkIndex {
+	return ips, nil
+}
+
+func (h *netlinkHandle) appendAddresses(ips sets.String, ifi *net.Interface) {
+	addr, err := ifi.Addrs()
+	if err != nil {
+		klog.Errorf("Can't get addresses from %s\n", ifi.Name)
+		return
+	}
+	for _, a := range addr {
+		var ip net.IP
+		switch v := a.(type) {
+		case *net.IPAddr:
+			ip = v.IP
+		case *net.IPNet:
+			ip = v.IP
+		default:
 			continue
 		}
 		if h.isIPv6 {
-			if route.Dst.IP.To4() == nil && !route.Dst.IP.IsLinkLocalUnicast() {
-				res.Insert(route.Dst.IP.String())
+			if ip.To4() != nil || ip.IsLinkLocalUnicast() {
```

Comment:
this applies to ipv4 to, do you want to discard them only for ipv6?
https://golang.org/src/net/ip.go?s=4256:4294#L143

Reply:
I keep the link-local check for ipv6 only. The thing is that for ipv6 a "scope" is required for link-local addresses, which is an interface, like `fe80::200:ff:fe01:1%eth0` and they are usually auto-assigned so they are always there. For ipv4 it would be possible to use them I think, and whey must have been configured by some one.